src/hotspot/share/runtime/crac.cpp

```diff
@@ -521,45 +476,62 @@ void crac::prepare_restore(crac_restore_data& restore_data) {
 }
 
 void crac::restore(crac_restore_data& restore_data) {
+  precond(CRaCRestoreFrom != nullptr);
+
   struct stat statbuf;
   if (os::stat(CRaCRestoreFrom, &statbuf) != 0) {
-    fprintf(stderr, "Cannot open restore directory of the -XX:CRaCRestoreFrom parameter: ");
-    perror(CRaCRestoreFrom);
+    log_error(crac)("Cannot open CRaCRestoreFrom=%s: %s", CRaCRestoreFrom, os::strerror(errno));
     return;
   }
   if ((statbuf.st_mode & S_IFMT) != S_IFDIR) {
-    fprintf(stderr, "-XX:CRaCRestoreFrom parameter is not a directory: %s\n", CRaCRestoreFrom);
+    log_error(crac)("CRaCRestoreFrom=%s is not a directory", CRaCRestoreFrom);
     return;
   }
 
-  compute_crengine();
-
-  const int id = os::current_process_id();
+  // Note that this is a local, i.e. the handle will be destroyed if we fail to restore
+  auto engine = CracEngine(CRaCRestoreFrom);
```

Comment:
A nitpick/question: why do you use this syntax that does an extra copy rather than
```
CracEngine engine(CRaCRestoreFrom);
```
?

Reply:
This is a mistake, thanks for catching